air/src/stack/field_ops/mod.rs

```diff
@@ -0,0 +1,190 @@
+    let mut constraint_offset = 0;
+
+    // Enforces the first element in the next frame is an inverse of the first element in the
+    // current frame.
+    result[constraint_offset] =
+        op_flag * are_equal(frame.selector(0) * frame.selector_next(0) - E::ONE, E::ZERO);
+
+    constraint_offset += 1;
+
+    constraint_offset
+}
+
+/// Enforces unique constraints of the NEG operation. The NEG operation updates the
+/// top element in the stack with its inverse. Therefore, the following constraints are enforced:
+/// - The next element in the next frame should be the negative of frist element in the
+///   current frame.
+pub fn enforce_neg_unique_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: E,
+) -> usize {
+    let mut constraint_offset = 0;
+
+    // Enforces the first element in the current frame is a negation of the first element in the
+    // next frame
+    result[constraint_offset] =
+        op_flag * are_equal(frame.selector(0) + frame.selector_next(0), E::ZERO);
+
+    constraint_offset += 1;
+
+    constraint_offset
+}
+
+/// Enforces unique constraints of the NOT operation. The NOT operation updates the
+/// top element in the stack with its bitwise not value. Therefore, the following constraints
+/// are enforced:
+/// - The top element in the stack should be a binary.
+/// - The first element of the next frame should be a binary not of the first element of
+/// the current frame.
+pub fn enforce_not_unique_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: E,
+) -> usize {
+    let mut constraint_offset = 0;
+
+    // Enforces the top element in the current frame is a binary or not.
+    result[constraint_offset] = op_flag * is_binary(frame.selector(0));
+
+    constraint_offset += 1;
+
+    // Enforces the first element in the current frame is a binary not of the first element
+    // in the next frame.
+    result[constraint_offset] =
+        op_flag * are_equal(frame.selector_next(0) + frame.selector(0) - E::ONE, E::ZERO);
+
+    constraint_offset += 1;
+
+    constraint_offset
+}
+
+trait EvaluationFrameExt<E: FieldElement> {
+    // --- Column accessors -----------------------------------------------------------------------
+
+    /// Gets the current element of the specified selector register in the stack.
+    fn selector(&self, index: usize) -> E;
+    /// Gets the next element of the specified selector register in the stack.
+    fn selector_next(&self, index: usize) -> E;
+}
+
+impl<E: FieldElement> EvaluationFrameExt<E> for &EvaluationFrame<E> {
+    // --- Column accessors -----------------------------------------------------------------------
+
+    #[inline(always)]
+    fn selector(&self, index: usize) -> E {
+        self.current()[STACK_TRACE_OFFSET + index]
+    }
+    #[inline(always)]
+    fn selector_next(&self, index: usize) -> E {
+        self.next()[STACK_TRACE_OFFSET + index]
+    }
```

Comment:
I think I had this comment during the previous review as well: why are we calling stack items "selectors"? These are just items on the stack, they are not selecting anything...